### PR TITLE
fix: objects in args

### PR DESCRIPTION
**Pull request by ssddOnTop**
Issue Reference(s):
Fixes https://github.com/tailcallhq/tailcall/issues/2999

### CODE REVIEW

File: src/core/blueprint/operators/http.rs [hunk 1]
```diff
@@ -1,18 +1,79 @@
...
+        return Scalar::is_predefined(cur_ty) || module.find_enum(cur_ty).is_some();
+    }
+    let type_ = type_.unwrap();
+
+    let cur = iter.next();
+    if cur.is_none() {
+        return Scalar::is_predefined(cur_ty) || module.find_enum(cur_ty).is_some();
+    }
+
+    let cur = cur.unwrap();
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:28]: Think of an alternative design where we don't use unwrap.

File: src/core/blueprint/operators/apollo_federation.rs [hunk 2]
```diff
@@ -36,8 +36,9 @@ pub fn compile_entity_resolver(inputs: CompileEntityResolver<'_>) -> Valid<IR, S
                 Resolver::Http(http) => compile_http(
                     config_module,
                     http,
+                    &Default::default(),
                     // inner resolver should resolve only single instance of type, not a list
-                    false,
```
Comment (tusharmath) [inline src/core/blueprint/operators/apollo_federation.rs:40]: We need to guarantee over here that the entity isn't a list. Why are we overriding it inside this the function?

Comment (ssddOnTop) [reply src/core/blueprint/operators/apollo_federation.rs:40]: I think there is a confusion.

Earlier the boolean was passed to check if the field `is_list`

but now I am passing the entire field and checking `field.type_of.is_list()`. In the case of federation, the boolean is replaced by `Field::default()` so `field.type_of.is_list()` will be false by default.

Now, I added the 4th parameter i.e. `is_federation` which is to skip the checks on args, as there are no args when http directive is defined on a Type.

File: src/core/blueprint/operators/http.rs [hunk 3]
```diff
@@ -30,19 +87,26 @@ pub fn compile_http(
...
+                let mut ans = Valid::succeed((
+                    key_value.key,
+                    key_value.value,
+                    key_value.skip_empty.unwrap_or_default(),
+                ));
+
+                // in the case when directive is defined on type
+                // we don't have any args to check for, so we can
+                // skip this check
+                if !is_federation && !check_scalar(&mustache, config_module, field) {
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:101]: Need to check for `value.a` where `a` is an object type.

File: src/core/blueprint/operators/http.rs [hunk 4]
```diff
@@ -1,18 +1,75 @@
...
+    }
+    ans
+}
+
 pub fn compile_http(
     config_module: &config::ConfigModule,
     http: &config::Http,
-    is_list: bool,
+    field: &Field,
+    is_federation: bool,
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:70]: We don't need `is_federation`

File: src/core/blueprint/operators/http.rs [hunk 5]
```diff
@@ -1,18 +1,75 @@
...
+}
+
+// Check if args are scalar
+fn check_args(mut iter: Iter<String>, module: &ConfigModule, field: &Field) -> bool {
+    let cur = iter.next();
+    if cur.is_none() {
+        return Scalar::is_predefined(field.type_of.name())
+            || module.find_enum(field.type_of.name()).is_some();
+    }
+    let cur = cur.unwrap();
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:39]: Handle these scenarios more gracefully.

File: src/core/blueprint/operators/http.rs [hunk 6]
```diff
@@ -1,18 +1,75 @@
...
+    field.args.contains_key(cur)
+        && check_ty(
+            iter,
+            module,
+            field.args.get(cur).as_ref().unwrap().type_of.name(),
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:44]: handle the unwrap more gracefully.

File: src/core/blueprint/operators/apollo_federation.rs [hunk 7]
```diff
@@ -37,7 +37,8 @@ pub fn compile_entity_resolver(inputs: CompileEntityResolver<'_>) -> Valid<IR, S
                     config_module,
                     http,
                     // inner resolver should resolve only single instance of type, not a list
-                    false,
+                    &Default::default(),
```
Comment (tusharmath) [inline src/core/blueprint/operators/apollo_federation.rs:40]: Pass Type and Args over here. Can create a new struct — `HttpParams {type_of: Option<Type>, args: Option<Args>}`  to set the HttpParams for the compiler to work.

**Pull request closed without merging**